web/src/message_events.js

```diff
@@ -118,21 +118,44 @@ export function update_views_filtered_on_message_property(
         // In case of multiple messages, we just rerender the view
         // since it is likely to use similar amount of resources to
         // fetching the messages and rerendering the view.
-        if (messages_to_fetch.length > 1 || !filter.can_apply_locally()) {
-            // TODO: Might be better to check if `/messages/matches_narrow`
-            // rather than doing a full rerender.
-            message_view.show(filter.terms(), {
-                then_select_id: msg_list.selected_id(),
-                then_select_offset: msg_list.selected_row().get_offset_to_window().top,
-                trigger: "message property update",
-                force_rerender: true,
+        if (!filter.can_apply_locally()) {
+            channel.get({
+                url: "/json/messages",
+                data: {
+                    message_ids: JSON.stringify(message_ids),
+                    narrow: JSON.stringify(filter.terms()),
+                },
+                success(data) {
+                    const messages_to_add = [];
+                    const messages_to_remove = new Set(message_ids);
+                    for (const raw_message of data.messages) {
+                        messages_to_remove.delete(raw_message.id);
+                        let message = message_store.get(raw_message.id);
+                        if (!message) {
+                            message = message_helper.process_new_message(raw_message);
+                        }
+                        messages_to_add.push(message);
+                    }
+                    msg_list.data.remove([...messages_to_remove]);
+                    msg_list.data.add_messages(messages_to_add);
+                    msg_list.rerender();
+                },
```

Comment:
The structure here, where we do this `channel.get` call inside the `message_lists.all_rendered_message_lists` loop, means we could in theory do multiple of this query in a way that's somewhat inefficient.

I also note that I think we need to do this logic for non-rendered cached message_list_data objects. So we might need, perhaps in a follow-up PR, to think about reworking the loop structure to first collect the messages to refetch into a single server query? Maybe I'm missing something.

Reply:
I agree we need to think about optimizing it but will be more to think about when adding the logic for cached message list data objects.